docs/contributing-testing.md

```diff
@@ -0,0 +1,106 @@
+# Testing
+
+This section describes how to run Wasmtime's tests and add new tests.
+
+Before continuing, make sure you can [build
+Wasmtime](./contributing-building.html) successfully. Can't run the tests if you
+can't build it!
+
+## Running All Tests
+
+To run all of Wasmtime's tests, execute this command:
+
+```shell
+cargo test --all
+```
+
+You can also exclude a particular crate from testing with `--exclude`. For
+example, if you want to avoid testing the `wastime-fuzzing` crate — which
+requires that `libclang` is installed on your system, and for some reason maybe
+you don't have it — you can run:
+
+```shell
+cargo test --all --exclude wasmtime-fuzzing
+```
+
+## Testing a Specific Crate
+
+You can test a particular Wasmtime crate with `cargo test -p
+wasmtime-whatever`. For example, to test the `wasmtime-environ` crate, execute
+this command:
+
+```shell
+cargo test -p wasmtime-environ
+```
+
+Alternatively, you can `cd` into the crate's directory, and run `cargo test`
+there, without needing to supply the `-p` flag:
+
+```shell
+cd crates/environ/
+cargo test
+```
+
+## Running the Wasm Spec Tests
+
+The spec testsuite itself is in a git submodule, so make sure you've
+checked it out and initialized its submodule:
+
+```shell
+git submodule update --init
+```
+
+When the submodule is checked out, Wasmtime runs the Wasm spec testsuite as part
+of testing the `wasmtime-cli` crate:
+
+```shell
+cargo test -p wasmtime-cli
+```
+
+## Adding New Tests
+
+### Adding Rust's `#[test]`-Style Tests
+
+For very "unit-y" tests, we add `test` modules in the same `.rs` file as the
+code that is being tested. These `test` modules are configured to only get
+compiled during testing with `#[cfg(test)]`.
+
+```rust
+// some code...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn some_test_for_that_code() {
+        // ...
+    }
+}
+```
+
+If you're writing a unit test and a `test` module doesn't already exist, you can
+create one.
+
+For more "integration-y" tests, we create a `tests` directory within the crate,
+and put the tests inside there. For example, there are various code
+cache-related tests at `crates/environ/tests/cache_*.rs`. Always feel free to
+add a `tests` directory to a crate, if you want to add a new test and there
+aren't any existing tests.
+
+### Adding Specification-Style Wast Tests
+
+We use the spec testsuite as-is and without custom patches or a forked
+version. If you have a new test that you want to add to it, make sure it makes
+sense for every Wasm implementation to run your test (i.e. it isn't
+Wasmtime-specific) and send a pull request
+[upstream](https://github.com/WebAssembly/testsuite/). Once it is accepted in
```

Comment:
I'm not actually sure how this repository is organized, but I think it's a scrape of all other repositories. I think the actual source for these tests lives in https://github.com/WebAssembly/spec/tree/master/test/core?

In any case I think it's probably best to suggest that adding to `misc_testsuite` is a good first step, and it should be on us as maintainers to periodically to move that to the upstream test suite. Primarily adding tests through the WebAssembly organization might just delay and or prevent test from being written because it would take longer.

Reply:
Agreed, I didn't mean to imply that upstreaming into the spec test suite was preferred, it's just hard to talk about the misc testsuite without talking about the spec tests. I'll wordsmith it a bit.